### ISSUE TITLE
clang: ICE in X86AsmParser when using incorrect operand number in inline assembly

### ISSUE THREAD

**Issue by AbyssStaror** [llvm/llvm-project#170111]
### Link for quick verification: https://godbolt.org/z/hMq3Yx85v
```c
#include <stdio.h>

// Inline assembly function that modifies array elements using asm volatile statements
void foo(int *b, int idx) {
    asm volatile (
        "movl $42, (%0)\n\t"   // write 42 to b[0]
        "movl $84, 4(%0)\n\t"  // write 84 to b[1]
        : "=m"(b[0]), "=m"(b[1])  // output operand to memory
        : "r"(b)                // use the eax to load b
        : "memory"
       );
}
```
Go back to `clang 8.0.0 (assertions)`.

### Smmary
When compiling the provided C++ code with clang, an internal compiler(ICE) error occurs due to an assertion failure in the X86AsmParser. The assertion `!isa<X86MCExpr>(Disp) && "Expected non-register here."` failed in `X86AsmParser.cpp`.
### Expected Behavior
The compiler should emit a clear error message indicating that the operand number %0is incorrectly used, as it references a memory output operand when a register input operand is expected for the memory addressing.

## Stack dump:
```
clang-21: /workspace/install/llvm/src/llvm-project/llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp:3044: bool {anonymous}::X86AsmParser::ParseMemOperand(llvm::MCRegister, const llvm::MCExpr*, llvm::SMLoc, llvm::SMLoc, llvm::OperandVector&): Assertion `!isa<X86MCExpr>(Disp) && "Expected non-register here."' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /workspace/install/llvm/build_21.1.4/bin/clang-21 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -dumpdir a- -disable-free -clear-ast-before-backend -main-file-name 38145.fuzz -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fdebug-compilation-dir=/workspace/exps -fcoverage-compilation-dir=/workspace/exps -resource-dir /workspace/install/llvm/build_21.1.4/lib/clang/21 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/x86_64-linux-gnu/c++/13 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/backward -internal-isystem /workspace/install/llvm/build_21.1.4/lib/clang/21/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -std=c++20 -fdeprecated-macro -ferror-limit 19 -fmessage-length=167 -fgnuc-version=4.2.1 -fno-implicit-modules -fskip-odr-check-in-gmf -fcxx-exceptions -fexceptions -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/38145-50e7e5.o -x c++ /workspace/exps/featurefuzz/generate_feature_fine_tuning_qwen_4b_choose_3_1_exp/cases/38145.fuzz
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module '/workspace/exps/featurefuzz/generate_feature_fine_tuning_qwen_4b_choose_3_1_exp/cases/38145.fuzz'.
4.      Running pass 'X86 Assembly Printer' on function '@_Z3fooPii'
 #0 0x0000617b98d16c2a llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x9691c2a)
 #1 0x0000617b98d17457 PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #2 0x0000617b98d10f57 llvm::sys::RunSignalHandlers() (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x968bf57)
 #3 0x0000617b98d111c7 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #4 0x0000793b502f6330 (/lib/x86_64-linux-gnu/libc.so.6+0x45330)
 #5 0x0000793b5034fb2c __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #6 0x0000793b5034fb2c __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #7 0x0000793b5034fb2c pthread_kill ./nptl/pthread_kill.c:89:10
 #8 0x0000793b502f627e raise ./signal/../sysdeps/posix/raise.c:27:6
 #9 0x0000793b502d98ff abort ./stdlib/abort.c:81:7
#10 0x0000793b502d981b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#11 0x0000793b502ec517 (/lib/x86_64-linux-gnu/libc.so.6+0x3b517)
#12 0x0000617b96ecb496 (anonymous namespace)::X86AsmParser::parseATTOperand(llvm::SmallVectorImpl<std::unique_ptr<llvm::MCParsedAsmOperand, std::default_delete<llvm::MCParsedAsmOperand>>>&) X86AsmParser.cpp:0:0
#13 0x0000617b96ece9d6 (anonymous namespace)::X86AsmParser::parseInstruction(llvm::ParseInstructionInfo&, llvm::StringRef, llvm::SMLoc, llvm::SmallVectorImpl<std::unique_ptr<llvm::MCParsedAsmOperand, std::default_delete<llvm::MCParsedAsmOperand>>>&) X86AsmParser.cpp:0:0
#14 0x0000617b9890d045 (anonymous namespace)::AsmParser::parseAndMatchAndEmitTargetInstruction((anonymous namespace)::ParseStatementInfo&, llvm::StringRef, llvm::AsmToken, llvm::SMLoc) AsmParser.cpp:0:0
#15 0x0000617b9892b9b4 (anonymous namespace)::AsmParser::parseStatement((anonymous namespace)::ParseStatementInfo&, llvm::MCAsmParserSemaCallback*) AsmParser.cpp:0:0
#16 0x0000617b989306c1 (anonymous namespace)::AsmParser::Run(bool, bool) AsmParser.cpp:0:0
#17 0x0000617b9a3d4f58 llvm::AsmPrinter::emitInlineAsm(llvm::StringRef, llvm::MCSubtargetInfo const&, llvm::MCTargetOptions const&, llvm::MDNode const*, llvm::InlineAsm::AsmDialect) const (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xad4ff58)
#18 0x0000617b9a3d5939 llvm::AsmPrinter::emitInlineAsm(llvm::MachineInstr const*) const (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xad50939)
#19 0x0000617b9a3c82b3 llvm::AsmPrinter::emitFunctionBody() (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xad432b3)
#20 0x0000617b968a6602 llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x7221602)
#21 0x0000617b97839586 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x81b4586)
#22 0x0000617b982336a9 llvm::FPPassManager::runOnFunction(llvm::Function&) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x8bae6a9)
#23 0x0000617b98233e89 llvm::FPPassManager::runOnModule(llvm::Module&) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x8baee89)
#24 0x0000617b98232b77 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x8badb77)
#25 0x0000617b98233469 llvm::legacy::PassManager::run(llvm::Module&) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x8bae469)
#26 0x0000617b9914249e clang::emitBackendOutput(clang::CompilerInstance&, clang::CodeGenOptions&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x9abd49e)
#27 0x0000617b99b2b3a2 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xa4a63a2)
#28 0x0000617b9c1f6eb5 clang::ParseAST(clang::Sema&, bool, bool) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xcb71eb5)
#29 0x0000617b99f3e92b clang::ASTFrontendAction::ExecuteAction() (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xa8b992b)
#30 0x0000617b99b2c360 clang::CodeGenAction::ExecuteAction() (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xa4a7360)
#31 0x0000617b99f469cf clang::FrontendAction::Execute() (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xa8c19cf)
#32 0x0000617b99e8cf46 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xa807f46)
#33 0x0000617b9a0cef7b clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0xaa49f7b)
#34 0x0000617b93e3d523 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x47b8523)
#35 0x0000617b93e2dc58 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#36 0x0000617b93e33b28 clang_main(int, char**, llvm::ToolContext const&) (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x47aeb28)
#37 0x0000617b93c8eb4a main (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x4609b4a)
#38 0x0000793b502db1ca __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#39 0x0000793b502db28b call_init ./csu/../csu/libc-start.c:128:20
#40 0x0000793b502db28b __libc_start_main ./csu/../csu/libc-start.c:347:5
#41 0x0000617b93e2bf55 _start (/workspace/install/llvm/build_21.1.4/bin/clang-21+0x47a6f55)
clang++: error: unable to execute command: Aborted (core dumped)
```
